flow/layers/display_list_raster_cache_item.cc

```diff
@@ -0,0 +1,188 @@
+  }
+
+  if (is_complex) {
+    // The caller seems to have extra information about the display list and
+    // thinks the display list is always worth rasterizing.
+    return true;
+  }
+
+  unsigned int complexity_score = complexity_calculator->Compute(display_list);
+  return complexity_calculator->ShouldBeCached(complexity_score);
+}
+
+DisplayListRasterCacheItem::DisplayListRasterCacheItem(
+    DisplayList* display_list,
+    const SkPoint& offset,
+    bool is_complex,
+    bool will_change)
+    : RasterCacheItem(RasterCacheKeyID(display_list->unique_id(),
+                                       RasterCacheKeyType::kDisplayList),
+                      CacheState::kCurrent),
+      display_list_(display_list),
+      offset_(offset),
+      is_complex_(is_complex),
+      will_change_(will_change) {}
+
+std::unique_ptr<DisplayListRasterCacheItem> DisplayListRasterCacheItem::Make(
+    DisplayList* display_list,
+    const SkPoint& offset,
+    bool is_complex,
+    bool will_change) {
+  return std::make_unique<DisplayListRasterCacheItem>(display_list, offset,
+                                                      is_complex, will_change);
+}
+
+void DisplayListRasterCacheItem::PrerollSetup(PrerollContext* context,
+                                              const SkMatrix& matrix) {
+  cache_state_ = CacheState::kNone;
+  DisplayListComplexityCalculator* complexity_calculator =
+      context->gr_context ? DisplayListComplexityCalculator::GetForBackend(
+                                context->gr_context->backend())
+                          : DisplayListComplexityCalculator::GetForSoftware();
+
+  if (!IsDisplayListWorthRasterizing(display_list_, will_change_, is_complex_,
+                                     complexity_calculator)) {
+    // We only deal with display lists that are worthy of rasterization.
+    return;
+  }
+
+  transformation_matrix_ = matrix;
+  transformation_matrix_.preTranslate(offset_.x(), offset_.y());
+
+  if (!transformation_matrix_.invert(nullptr)) {
+    // The matrix was singular. No point in going further.
+    return;
+  }
+
+  if (context->raster_cached_entries && context->raster_cache) {
+    context->raster_cached_entries->push_back(this);
+    cache_state_ = CacheState::kCurrent;
+  }
+  return;
+}
+
+void DisplayListRasterCacheItem::PrerollFinalize(PrerollContext* context,
+                                                 const SkMatrix& matrix) {
+  if (cache_state_ == CacheState::kNone || !context->raster_cache ||
+      !context->raster_cached_entries) {
+    return;
+  }
+  auto* raster_cache = context->raster_cache;
+  SkRect bounds = display_list_->bounds().makeOffset(offset_.x(), offset_.y());
+  // We've marked the cache entry that we would like to cache so it stays
+  // alive, but if the following conditions apply then we need to set our
+  // state back to !cacheable_ so that we don't populate the entry later.
+  if (!SkRect::Intersects(context->cull_rect, bounds)) {
```

Comment:
Aboud this do you mean we can 
```C++
void DisplayListRasterCacheItem::PrerollFinalize(PrerollContext* context,
                                                 const SkMatrix& matrix) {
  if (cache_state_ == CacheState::kNone || !context->raster_cache ||
      !context->raster_cached_entries) {
    return;
  }
  auto* raster_cache = context->raster_cache;
  SkRect bounds = display_list_->bounds().makeOffset(offset_.x(), offset_.y());
  if (!SkRect::Intersects(context->cull_rect, bounds) || raster_cache->MarkSeen(key_id_, transformation_matrix_) <
      raster_cache->access_threshold()) {
    cache_state_ = CacheState::kNone;
    return;
  }
  context->subtree_can_inherit_opacity = true;
  cache_state_ = CacheState::kCurrent;
  return;
}
```
Like this?

Reply:
Perhaps do the tests the other way around so MarkSeen continues to mark it as seen even if it is not in the cull rect.

Reply:
That means we create display_list Entry whether the rectangles intersect or not

Reply:
True, but the alternative is to let it die. Look at the old code: https://github.com/flutter/engine/blob/1d4fe46c5da80122954948229d32b1af44491693/flow/layers/display_list_layer.cc#L100

Before we would "Touch" the entry if it didn't intersect which would keep it alive (if it existed). Looking at the implementation of "Touch", it wouldn't create an entry, but it would prevent it from going away. That may be a logic flaw because "Touch" would definitely aid in the "access_count" countdown so it is not clear why it wouldn't also create the entry.

Without calling MarkSeen, though, your code will allow it to die as soon as it leaves the clip as it will never have its "used this frame" flag set.

I don't think it is worth creating a new method/flow in the new mechanism to "MarkSeenIfCacheEntryExists" as I think it is likely a bug that the old code waited until the item intersected the clip to make it start the countdown. (If the intent was to not start the countdown, then why increment access_count at all in the old Touch method?)

Reply:
I chatted with the engineer who added the access_count++ line to the Touch method and apparently this was added to deal with partial repaints. When an item scrolls on screen, but then is excluded from subsequent frames due to a tiny damage rect, the intent was to keep incrementing access_count in that case so that it keeps accumulating frames just as if the partial repaint didn't exist.

So, to maintain precise compatibility here, MarkSeen (or a similar method) would need to avoid creating the entry if the bounds are culled.

We should not let an existing entry die, though, in this culling case as that would create bigger problems with partial repaint flushing everything out of the cache.

Reply:
Okey, In this case i try use the old logic like this:
```C++

void DisplayListRasterCacheItem::PrerollFinalize(PrerollContext* context,
                                                 const SkMatrix& matrix) {
  ...
  // We must to create an entry whenever if the react is intersect.
  // if the rect is intersect we will get the entry access_count to confirm if it great than the threshold.
  // Otherwise we only increase the entry access_count.
  if (context->cull_rect.intersect(bounds)) {
    if (raster_cache->MarkSeen(key_id_, transformation_matrix_) <
        raster_cache->access_threshold()) {
      cache_state_ = CacheState::kNone;
      return;
    }
    context->subtree_can_inherit_opacity = true;
    cache_state_ = CacheState::kCurrent;
  } else {
    raster_cache->Touch(key_id_, matrix);
  }
  return;
}
```

Reply:
That looks good. I didn't realize that Touch still existed.

BTW, why do you use Touch in DLRCItem::Draw()? The entry should have already been touched by the time we try to draw it.

Reply:
If the display_list is should not cache itself, in the draw we will increase the access count.
The MarkSceen only create the entry and set use_this_frame = true, doesn't increase the access_count

Reply:
This logic is tested by : `RasterCache::MetricsOmitUnpopulatedEntries` test case

Reply:
The new changes to PrerollFinalize look OK now. I see what Touch is doing in Draw now. It is achieving the desired goal, but we should revisit how it does that after this PR lands.